File: problems/data-structures-n-algorithms/wip/937.cpp
// 937. Reorder Data in Log Files
// https://leetcode.com/problems/reorder-data-in-log-files/

#include <bits/stdc++.h>
#include <deque>
#include <iostream>
#include <vector>
#include <utility>
#include <cstdlib>
#include <locale>

using namespace std;

#include "../../../utils/customString.h"

// overall correct soln but does't work in c++ cuz of nature of sort function

int getWeight(char c) {
	int a;

	if (c > 57) {
		a = c - 'a' + 11;
	} else if (c >= 48) {
		a = c - '0' + 1;
	} else {
		a = 0;
	}

	return a;
}

bool isDigit(char c) {
	if (c >= 48 && c <= 57) {
		return true;
	}
	return false;
}

bool isChar(char c) {
	if (c >= 97 && c < 123) {
		return true;
	}
	return false;
}

// bool comp(string &a, string &b) {
// 	if (isDigit(a[a.size() - 1]) && isChar(b[b.size() - 1])) {
// 		return false;
// 	} else if (isChar(a[a.size() - 1]) && isDigit(b[b.size() - 1])) {
// 		return true;
// 	} else if (isDigit(a[a.size() - 1]) && isDigit(b[b.size() - 1])) {
// 		return false;
// 	}

// 	int ia = 0;
// 	int ib = 0;

// 	while (ia < a.size() && a[ia] != ' ') {
// 		ia++;
// 	}

// 	while (ib < b.size() && b[ib] != ' ') {
// 		ib++;
// 	}

// 	while (ia < a.size() && ib < b.size() && a[ia] == b[ib]) {
// 		ia++;
// 		ib++;
// 	}

// 	if (ia < a.size() && ib < b.size()) {
// 		return getWeight(a[ia]) < getWeight(b[ib]);
// 	} else if (ia == a.size() && ib < b.size()) {
// 		return true;
// 	} else if (ib == b.size() && ia < a.size()) {
// 		return false;
// 	}
// 	ia = 0;
// 	ib = 0;
// 	while (a[ia] != ' ' && b[ib] != ' ' && a[ia] == b[ib]) {
// 		ia++;
// 		ib++;
// 	}
// 	return getWeight(a[ia]) < getWeight(b[ib]);
// }

bool comp(string &a, string &b) {
	// if (isDigit(a[a.size() - 1])) {
	// 	return false;
	// } else if (isDigit(b[b.size() - 1])) {
	// 	return true;
	// }
	if (isDigit(a[a.size() - 1]) && isChar(b[b.size() - 1])) {
		return false;
	} else if (isChar(a[a.size() - 1]) && isDigit(b[b.size() - 1])) {
		return true;
	} else if (isDigit(a[a.size() - 1]) && isDigit(b[b.size() - 1])) {
		return false;
	}

	int ia = 0;
	int ib = 0;

	while (ia < a.size() && a[ia] != ' ') {
		ia++;
	}

	while (ib < b.size() && b[ib] != ' ') {
		ib++;
	}

	while (ia < a.size() && ib < b.size() && a[ia] == b[ib]) {
		ia++;
		ib++;
	}

	if (ia < a.size() && ib < b.size()) {
		return getWeight(a[ia]) <= getWeight(b[ib]);
	} else if (ia == a.size() && ib < b.size()) {
		return true;
	} else if (ib == b.size() && ia < a.size()) {
		return false;
	}
	ia = 0;
	ib = 0;
	while (a[ia] != ' ' && b[ib] != ' ' && a[ia] == b[ib]) {
		ia++;
		ib++;
	}
	return getWeight(a[ia]) <= getWeight(b[ib]);
}

vector<string> reorderLogFiles(vector<string> &logs) {
	//

	sort(logs.begin(), logs.end(), comp);

	return logs;
}

int main() {
	//

	vector<string> logs;

	logs = {"dig1 8 1 5 1", "let1 art can", "dig2 3 6", "let2 own kit dig",
					"let3 art zero"};

	vector<string> res = reorderLogFiles(logs);

	for (int i = 0; i < res.size(); i++) {
		cout << res[i] << ", ";
	}
	cout << endl;

	return 0;
}

File: problems/data-structures-n-algorithms/wip/937.java
// 937. Reorder Data in Log Files
// https://leetcode.com/problems/reorder-data-in-log-files/

import java.util.Arrays;
import java.util.Comparator;

class S937 {

	public static void main (String[] args) {
		String[] arr;
		// arr = new String[]{
		// 	"dig1 8 1 5 1", "let1 art can", "dig2 3 6", "let2 own kit dig", "let3 art zero"
		// };
		arr = new String[]{
			"a1 9 2 3 1", "g1 act car", "zo4 4 7", "ab1 off key dog", "a8 act zoo", "a2 act car"
		};

		String[] res = reorderLogFiles(arr);

		for (String s : res) {
			System.out.println(s);
		}
	}

	public static String[] reorderLogFiles (String[] logs) {
		Arrays.sort(logs, new Compare());
		return logs;
	}

	static class Compare implements Comparator<String> {
		public int compare (String a, String b) {
			if (Character.isDigit(a.charAt(a.length() - 1)) && Character.isAlphabetic(b.charAt(b.length() - 1))) {
				return 1;
			} else if (Character.isAlphabetic(a.charAt(a.length() - 1)) && Character.isDigit(b.charAt(b.length() - 1))) {
				return -1;
			} else if (Character.isDigit(a.charAt(a.length() - 1)) && Character.isDigit(b.charAt(b.length() - 1))) {
				return 0;
			}
			int ia = 0;
			int ib = 0;
			while (ia < a.length() && a.charAt(ia) != ' ') {
				ia++;
			}
			while (ib < b.length() && b.charAt(ib) != ' ') {
				ib++;
			}
			while (ia < a.length() && ib < b.length() && a.charAt(ia) == b.charAt(ib)) {
				ia++;
				ib++;
			}

			if (ia == a.length() && ib < b.length()) {
				return -1;
			} else if (ib == b.length() && ia < a.length()) {
				return 1;
			} else if (ia < a.length() && ib < b.length()) {
				return a.charAt(ia) - b.charAt(ib);
			}

			ia = 0;
			ib = 0;
			while (a.charAt(ia) != ' ' && b.charAt(ib) != ' ' && a.charAt(ia) == b.charAt(ib)) {
				ia++;
				ib++;
			}
			return a.charAt(ia) - b.charAt(ib);
		}
	}
}
